base/compiler/optimize.jl

```diff
@@ -1104,6 +1104,19 @@ function convert_to_ircode(ci::CodeInfo, sv::OptimizationState)
                         code[i] = nothing
                     end
                 end
+            elseif isa(expr, PhiNode)
+                new_edges = Int32[]
+                new_vals = Any[]
+                for j = 1:length(expr.edges)
+                    (expr.edges[j] in sv.unreachable) && continue
```

Comment:
```suggestion
                    (expr.edges[j] in sv.unreachable || ssavaluetypes[expr.edges[j]] === Union{}) && continue
```
Ah, it turns out `expr.edges[j] in sv.unreachable` is not enough, since must-throw statement can be a fall-through terminator.

Reply:
Good catch

Reply:
Later in the function we spell this as `ssavaluetypes[idx] === Union{} && !(oldidx in sv.unreachable) && !isa(code[idx], PhiNode)`, which suggests this probably needs the `PhiNode` case as well, or rather we should factor out the query and use it in both places.

Reply:
`is_must_throw(idx::Int) = ssavaluetypes[idx] === Union{}`?

Reply:
Well, I was suggesting factoring out all three conditions, but it's slightly awkward since the use later in the function is also renumbering things, so there's multiple index sets. I guess for now maybe just add the `PhiNode` condition here (and ideally a test for it).

Reply:
Hmm, what does 'The `PhiNode` condition' mean? At L1159, `!isa(code[idx], PhiNode)` is used to check if the statement being rewritten is a `PhiNode`. In this context, the statement being rewritten is `PhiNode` always, and I think the logic (`(edge in sv.unreachable || is_must_throw(edge))`) still applies even when `code[edge] isa PhiNode`.

Reply:
Understood, thanks for the explanation. I have no objections to putting in that check, but I found it challenging to come up with a test case. It seems to be a pretty rare scenario, so maybe it's alright to go ahead without it.